system/jlib/jstats.cpp

```diff
@@ -2091,7 +2091,35 @@ class TreeNodeStatisticGatherer : public CInterfaceOf<IStatisticGatherer>
 
 void CRuntimeStatistic::merge(unsigned __int64 otherValue, StatsMergeAction mergeAction)
 {
-    value = mergeStatisticValue(value, otherValue, mergeAction);
+    switch (mergeAction)
+    {
+    case StatsMergeKeepNonZero:
+        if (otherValue)
+        {
+            unsigned __int64 zero = 0;
+            value.compare_exchange_weak(zero, otherValue);
```

Comment:
this means that if otherValue is non zero, but something else has already set value, this will fail, i.e. the new otherValue will be ignored.

Prior to this change, sequential attempts to merge with StatsMergeKeepNonZero would succeed, and result in latest value being kept.

I think this should just set value (if otherValue is non-zero), to ensure latest is kept.

Reply:
@jakesmith Originally in mergeStatisticValue, it had: 
    case StatsMergeKeepNonZero:
        if (prevValue)
            return prevValue;
        return newValue;

So, it only updated to the newValue if the prevValue was zero.

Reply:
Ah yes, ok. 
'StatsMergeKeepNonZero' sounds like it will keep the new value if non-zero..

@ghalliday - is it correct to ignore new value with this mode?

Reply:
Yes, I think that is correct.